src/main/java/org/ibp/api/java/impl/middleware/germplasm/GermplasmCodeGenerationServiceImpl.java

```diff
@@ -0,0 +1,202 @@
+package org.ibp.api.java.impl.middleware.germplasm;
+
+import org.generationcp.commons.ruleengine.RuleException;
+import org.generationcp.commons.ruleengine.RuleFactory;
+import org.generationcp.commons.ruleengine.coding.CodingRuleExecutionContext;
+import org.generationcp.commons.ruleengine.service.RulesService;
+import org.generationcp.commons.service.GermplasmNamingService;
+import org.generationcp.middleware.api.germplasm.GermplasmService;
+import org.generationcp.middleware.api.nametype.GermplasmNameTypeDTO;
+import org.generationcp.middleware.api.nametype.GermplasmNameTypeService;
+import org.generationcp.middleware.domain.germplasm.GermplasmCodeNameBatchRequestDto;
+import org.generationcp.middleware.exceptions.InvalidGermplasmNameSettingException;
+import org.generationcp.middleware.pojos.Germplasm;
+import org.generationcp.middleware.pojos.Name;
+import org.generationcp.middleware.pojos.germplasm.GermplasmNameSetting;
+import org.generationcp.middleware.pojos.naming.NamingConfiguration;
+import org.generationcp.middleware.service.api.GermplasmCodingResult;
+import org.generationcp.middleware.service.api.GermplasmGroupingService;
+import org.generationcp.middleware.service.api.NamingConfigurationService;
+import org.generationcp.middleware.util.Util;
+import org.ibp.api.exception.ApiRuntimeException;
+import org.ibp.api.java.germplasm.GermplasmCodeGenerationService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Service to generate Code Names (aka. Group Names)
+ */
+@Transactional
+public class GermplasmCodeGenerationServiceImpl implements GermplasmCodeGenerationService {
```

Comment:
@abatac @darla-leafnode 
Seems to me you are coding like BMSPI is Middleware and I assume it is because some services you need to use are in Commons instead of Middleware (they were in Fieldbook in the past) 
Please check your code and see if there is any option for you to move whatever is related to save the names to Middleware or use DTO's but if you can move it to Middleware at least yo avoid mapping your objects
Let's see if we can move some code from Commons to Middleware in the near future :(

Reply:
HI @clarysabel - yes this service was previously in Commons and I asked Aldrin yesterday to move to BMSAPI and improve the code  (replacing code that accesses Middleware DAO and Hibernate entities). He'll continue to improve today. 
Afterwards, this service is a first step of what you said:
> Let's see if we can move some code from Commons to Middleware in the near future